core/profiles/src/main/java/software/amazon/awssdk/profiles/internal/ProfileFileRefresher.java

```diff
@@ -0,0 +1,267 @@
+import software.amazon.awssdk.annotations.SdkInternalApi;
+import software.amazon.awssdk.annotations.SdkTestInternalApi;
+import software.amazon.awssdk.profiles.ProfileFile;
+import software.amazon.awssdk.utils.SdkAutoCloseable;
+import software.amazon.awssdk.utils.cache.CachedSupplier;
+import software.amazon.awssdk.utils.cache.RefreshResult;
+
+/**
+ * Class used for caching and reloading ProfileFile objects from a Supplier.
+ */
+@SdkInternalApi
+public final class ProfileFileRefresher implements SdkAutoCloseable {
+
+    private static final ProfileFileRefreshRecord EMPTY_REFRESH_RECORD = ProfileFileRefreshRecord.builder()
+                                                                                                 .refreshTime(Instant.MIN)
+                                                                                                 .build();
+    private final CachedSupplier<ProfileFileRefreshRecord> profileFileCache;
+    private volatile ProfileFileRefreshRecord currentRefreshRecord;
+    private final Supplier<ProfileFile> profileFile;
+    private final Path profileFilePath;
+    private final Function<RuntimeException, ProfileFile> exceptionHandler;
+    private final Consumer<ProfileFile> onProfileFileReload;
+    private final Clock clock;
+
+    private ProfileFileRefresher(Builder builder) {
+        this.exceptionHandler = builder.exceptionHandler;
+        this.clock = builder.clock;
+        this.profileFile = builder.profileFile;
+        this.profileFilePath = builder.profileFilePath;
+        this.onProfileFileReload = builder.onProfileFileReload;
+        this.profileFileCache = CachedSupplier.builder(this::refreshResult)
+                                              .clock(this.clock)
+                                              .build();
+        this.currentRefreshRecord = EMPTY_REFRESH_RECORD;
+    }
+
+    /**
+     * Builder method to construct instance of ProfileFileRefresher.
+     */
+    public static ProfileFileRefresher.Builder builder() {
+        return new ProfileFileRefresher.Builder();
+    }
+
+    /**
+     * Retrieves the cache value or refreshes it if stale.
+     */
+    public ProfileFile refreshIfStale() {
+        ProfileFileRefreshRecord cachedOrRefreshedRecord = profileFileCache.get();
+        ProfileFile cachedOrRefreshedProfileFile = cachedOrRefreshedRecord.profileFile;
+        if (isNewProfileFile(cachedOrRefreshedProfileFile)) {
+            currentRefreshRecord = cachedOrRefreshedRecord;
+        }
+
+        return cachedOrRefreshedProfileFile;
+    }
+
+    @Override
+    public void close() {
+        profileFileCache.close();
+    }
+
+    private RefreshResult<ProfileFileRefreshRecord> refreshResult() {
+        try {
+            return reloadAsRefreshResultIfStale();
+        } catch (RuntimeException exception) {
+            Instant now = Instant.now();
+            Instant staleTime = now;
+            ProfileFile exceptionProfileFile = exceptionHandler.apply(exception);
```

Comment:
Why do we not just throw the exception? What's the purpose of an exception profile file?

Reply:
I was copying `CachedTokenRefresher`. I guess we could recover from an exception and provide a fallback profile file, maybe the default.

Reply:
Interesting. I was thinking the intuitive behavior would just be to throw the exception (failing whatever profile file refresh is happening) and then let the next get() call try again. Essentially, push the failure handling upstream.

Trying to do something more intelligent might be confusing for the user to understand.

Reply:
Agreed this may be too complex. Removevd.